Added parsing support for the const type

lukeg101 opened the pull request:
Consider the following fragment of a litmus C test:
```
P0 (const int* y,volatile int* x) {
  int r0 = *x;
  atomic_thread_fence(memory_order_release);
  *y = 1;
}
```
which uses `const` qualified types. This patch adds parsing support for
`const` qualified types. Like `volatile` we do not implement semantics
for `const` in Herd.

Additionally we add testing infrastructure to automatically test litmus
C tests against the `rc11.cat` memory model, like we do with `AArch64`.
As such, unittests are provided for const and volatile qualified types in herd

Following herd#33

lukeg101 committed:
Added parsing support for the const type

Consider the following fragment of a litmus C test:
```
P0 (const int* y,volatile int* x) {
  int r0 = *x;
  atomic_thread_fence(memory_order_release);
  *y = 1;
}
```
which uses `const` qualified types. This patch adds parsing support for
`const` qualified types. Like `volatile` we do not implement semantics
for `const` in Herd.

Additionally we add testing infrastructure to automatically test litmus
C tests against the `rc11.cat` memory model, like we do with `AArch64`.
As such, unittests are provided for const and volatile qualified types in herd

Following #33

relokin commented on Makefile:
```diff
@@ -108,6 +108,16 @@ test::
+		-litmus-dir ./herd/tests/instructions/C \
+		-conf ./herd/tests/instructions/C/c.cfg \
+		$(REGRESSION_TEST_MODE)
+	@ echo "herd7 AArch64 NEON instructions tests: OK"
```
These are not AArch64 NEON tests :)

lukeg101 replied:
haha that'll teach me

relokin commented on lib/CParser.mly:
```diff
@@ -163,10 +165,10 @@ expr:
-| CONSTVAR { Const(mk_sym $1) }
-| IDENTIFIER { LoadReg $1 }
-| STAR IDENTIFIER { LoadMem (LoadReg $2,AN []) }
+| CONSTANT { CBase.Const(Constant.Concrete $1) }
```
Fine by me but they shouldn't be necessary because of the `open CBase` statement in the prelude.

lukeg101 replied:
We do need these as herd will fail to build on older versions of Ocaml (<=4.00)

error:
```
File "lib/CParser.mly", line 169, characters 13-18:
Warning 42: this use of Const relies on type-directed disambiguation,
it will not compile with OCaml 4.00 or earlier.
```
It's the same reason everything in lib/AArch64Parser now uses disambiguation

relokin commented on lib/CType.ml:
```diff
@@ -100,37 +103,45 @@ let rec is_pte t = match t with
-let strip_volatile0 = function
-  | Volatile t -> t
-  | t -> t
+  | Atomic t -> strip_atomic t (* We handle the case where we have malformed types *)
```
I am not sure I understand the comment. Also this changes what strip_atomic does, `strip_atomic (Atomic t)` used to return t now it will return strip_atomic t. In most cases that should be fine but I don't see why we want to do it this way.

lukeg101 replied:
I think my reasoning was if _somehow_ a user provides a nested type like `Atomic (Atomic t)` we would catch it or `Volatile (Volatile t)`

lukeg101 replied:
I guess this is what I meant by malformed

relokin replied:
I think this should never happen. We shouldn't ever create a type Atomic (Atomic t) or Volatile (Volatile t), ideally we should catch these as parser errors. I think this for "compound" types like Const (Array t), Volatile (Pointer t) which make sense in C/C++.

relokin commented on lib/CType.ml:
```diff
@@ -100,37 +103,45 @@ let rec is_pte t = match t with
-  | t -> t
+  | Atomic t -> strip_atomic t (* We handle the case where we have malformed types *)
+  | Const t -> Const (strip_atomic t)
+  | Pointer t -> Pointer (strip_atomic t)
```
again this changes the behavior of `strip_volatile` I see why we may want to do it but I am not sure it will ever be used.

lukeg101 replied:
do you feel strongly about this? I am happy to remove it, but this way we get a guarantee that volatile is indeed removed in the pathological case. since CType is an open library I think we should assume the user might accidentally give the function a poorly formed type

relokin replied:
No let's take this conversation upstream.

relokin commented on lib/CType.ml:
```diff
@@ -100,37 +103,45 @@ let rec is_pte t = match t with
   | Pointer t -> Pointer (strip_volatile t)
-  | t ->  strip_volatile0 t
+  | Const t -> Const (strip_volatile t)
+  | Volatile t -> strip_volatile t (* We handle the case where we have malformed types*)
```
what are malformed types, maybe that will help me understand.

lukeg101 replied:
I'm not sure... I think this was an artifact of an old version of herd

relokin commented on herd/tests/instructions/C/LB000.litmus:
```diff
@@ -0,0 +1,25 @@
+C LB000
```
If our intention is to test `const`, I think `LB` is a very complex litmus test. Would it make sense to add a litmus test with a single write to a `const` location? We can also add other tests that aim that assert expectations from `stripe_const` and other `stripe_*` functions that change in this PR (and tbh, weren't clear even before this PR).

lukeg101 replied:
I am just adding parsing support in this patch, I would be happy to add source semantics support for const too, but I would need to add a new cat file with the const semantics like I implemented for aarch64. Patch starts to runaway here

relokin replied:
Good point and no need to add a cat file. In which case, I think the test that demonstrates what currently works is a litmus with a single load, maybe?

`LB`, apart from being complex and testing many more things (than the parsing of `const`), also demonstrates the lack of semantics for `const` as LB with `const` annotation shows the same executions as LB without `const` annotations.

lukeg101 replied:
yeah I agree, I'll put together a simple load test

lukeg101 commented:
Hi Nikos,
Thanks for taking a look at this, I did the following:
- fixed typo in Makefile
- did not remove type disambiguation of const (needed for backwards-compat)
- replaced LB tests with simple load tests
- did not change strip_volatile/atomic, but can do so if you think it should be removed.

relokin commented:
> Hi Nikos,
> Thanks for taking a look at this, I did the following:
> 
>     * fixed typo in Makefile
> 
>     * did not remove type disambiguation of const (needed for backwards-compat)
> 
>     * replaced LB tests with simple load tests
> 
>     * did not change strip_volatile/atomic, but can do so if you think it should be removed.

Thank you for the changes! Feel free to post this upstream and we can continue the review process there.
